### ISSUE TITLE
completion window prevents getline() from updating

### ISSUE THREAD

**Issue by TheZoq2** [neovim/neovim#7443]
- `nvim --version`:
```
NVIM v0.2.0
Build type: Release
Compilation: /usr/bin/cc -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -Wconversion -DNVIM_MSGPACK_HAS_FLOAT32 -O2 -DNDEBUG -DDISABLE_LOG -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -I/build/neovim/src/build/config -I/build/neovim/src/neovim-0.2.0/src -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/build/neovim/src/build/src/nvim/auto -I/build/neovim/src/build/include
Compiled by builduser

Optional features included (+) or not (-): +acl   +iconv    +jemalloc +tui      
For differences from Vim, see :help vim-differences

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"
```

- Vim version, works as it should
```
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Oct  7 2017 13:52:29)
Included patches: 1-1176
Compiled by Arch Linux
Huge version without GUI.  Features included (+) or not (-):
+acl             +file_in_path    +mouse_sgr       +tag_old_static
+arabic          +find_in_path    -mouse_sysmouse  -tag_any_white
+autocmd         +float           +mouse_urxvt     +tcl/dyn
-balloon_eval    +folding         +mouse_xterm     +termguicolors
-browse          -footer          +multi_byte      +terminal
++builtin_terms  +fork()          +multi_lang      +terminfo
+byte_offset     +gettext         -mzscheme        +termresponse
+channel         -hangul_input    +netbeans_intg   +textobjects
+cindent         +iconv           +num64           +timers
-clientserver    +insert_expand   +packages        +title
-clipboard       +job             +path_extra      -toolbar
+cmdline_compl   +jumplist        +perl/dyn        +user_commands
+cmdline_hist    +keymap          +persistent_undo +vertsplit
+cmdline_info    +lambda          +postscript      +virtualedit
+comments        +langmap         +printer         +visual
+conceal         +libcall         +profile         +visualextra
+cryptv          +linebreak       +python/dyn      +viminfo
+cscope          +lispindent      +python3/dyn     +vreplace
+cursorbind      +listcmds        +quickfix        +wildignore
+cursorshape     +localmap        +reltime         +wildmenu
+dialog_con      +lua/dyn         +rightleft       +windows
+diff            +menu            +ruby/dyn        +writebackup
+digraphs        +mksession       +scrollbind      -X11
-dnd             +modify_fname    +signs           -xfontset
-ebcdic          +mouse           +smartindent     -xim
+emacs_tags      -mouseshape      +startuptime     -xpm
+eval            +mouse_dec       +statusline      -xsmp
+ex_extra        +mouse_gpm       -sun_workshop    -xterm_clipboard
+extra_search    -mouse_jsbterm   +syntax          -xterm_save
+farsi           +mouse_netterm   +tag_binary      
   system vimrc file: "/etc/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
  fall-back for $VIM: "/usr/share/vim"
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H   -D_FORTIFY_SOURCE=2  -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fno-plt -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1       
Linking: gcc   -L. -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -fstack-protector -rdynamic -Wl,-export-dynamic -Wl,-E -Wl,-rpath,/usr/lib/perl5/5.26/core_perl/CORE  -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -L/usr/local/lib -Wl,--as-needed -o vim        -lm -lncurses -lelf -lnsl    -lacl -lattr -lgpm -ldl   -Wl,-E -Wl,-rpath,/usr/lib/perl5/5.26/core_perl/CORE -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -fstack-protector-strong -L/usr/local/lib  -L/usr/lib/perl5/5.26/core_perl/CORE -lperl -lpthread -lnsl -ldl -lm -lcrypt -lutil -lc   -L/usr/lib -ltclstub8.6 -ldl -lz -lpthread -lieee -lm
```
- Operating system/version:
Archlinux, recently updated
- Terminal name/version:
xfce4-terminal
- `$TERM`:
xterm-256color

### Steps to reproduce
Use the following minimal vim rc

```viml
set nocompatible

filetype indent plugin on

" Hide messages on status line.
set noshowmode
set shortmess+=c

" Trigger TestComplete completions at each keypress then print the current line.
autocmd TextChangedI * call feedkeys("\<C-X>\<C-U>\<C-P>") | echom "Current line: " . getline(".")

function! TestComplete(findstart, base)
    if a:findstart
        return col('.')
    endif
    " Refresh always is needed to trigger the TextChangedI event once the completion menu is open.
    return { 'words': ['test1', 'test2'], 'refresh': 'always' }
endfunction

set completefunc=TestComplete
```

### Actual behaviour
If I type `test `, the "Current line: " print message is empty until I press space. Using backspace updates it but adding new letters doesn't update it until space or backspace is pressed.

### Expected behaviour
The printed "current line: " message should be updated for each character.

This is a problem when using the autoexpand feature of ultisnips and I originally reported the issue on the youcompleteme repo thinking it was a plugin issue. https://github.com/Valloric/YouCompleteMe/issues/2524

**Comment by justinmk** [comment]
Thanks for the report. Can you try the development version? The [Releases](https://github.com/neovim/neovim/releases) page has pre-built archives for Linux/Windows/macOS. (No installation needed, just unzip and run. For Linux, [nvim.appimage](https://github.com/neovim/neovim/releases/download/nightly/nvim.appimage) is a universal "binary", just `chmod u+x nvim.appimage && ./nvim.appimage`)

**Comment by Shougo** [comment]
> If I type test, the "Current line: " print message is empty until I press space. Using backspace updates it but adding new letters doesn't update it until space or backspace is pressed.

It seems the feature.
Because the candidate is not inserted until the close the popup window.

**Comment by TheZoq2** [comment]
I tested using the master branch and the issue persists. Here is the output of `nvim --version`

```
NVIM v0.2.1-1060-gf0c2f82e9
Build type: RelWithDebInfo
LuaJIT 2.0.5
Compilation: /usr/bin/cc -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fno-plt -Wconversion -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -O2 -g -DMIN_LOG_LEVEL=3 -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -I/tmp/yaourt-tmp-frans/aur-neovim-git/src/neovim-git/build/config -I/tmp/yaourt-tmp-frans/aur-neovim-git/src/neovim-git/src -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/tmp/yaourt-tmp-frans/aur-neovim-git/src/neovim-git/build/src/nvim/auto -I/tmp/yaourt-tmp-frans/aur-neovim-git/src/neovim-git/build/include
Compiled by frans@frans-arch

Features: +acl +iconv +jemalloc +tui 
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

> It seems the feature.
> Because the candidate is not inserted until the close the popup window.

No, I never insert any of the candidates. Whenever suggestions are being shown, the content of the getline output stops updating when im typing normally.

**Comment by mljli** [comment]
> If I type test, the "Current line: " print message is empty until I press space. Using backspace updates it but adding new letters doesn't update it until space or backspace is pressed.

All the punctuation marks except the underscore also work.

**Comment by TheZoq2** [comment]
I started looking into this and it seems like the problem is that the `TextChangedI` autocmd does not get called when the completion window is open.

At least if my way of testing is correct. I ran vim, attached `gdb` to it and set a breakpoint in `f_getline`. The breakpoint is only triggered when the completion window is closed

**Comment by chrisbra** [comment]
> I started looking into this and it seems like the problem is that the TextChangedI autocmd does not get called when the completion window is open.

Correct, at least for Vim:
https://github.com/vim/vim/blob/0479e910c423d71e2b96bc721feffad5808e767a/src/edit.c#L1677-L1686

There is vim/vim#2372 to add TextChangedP autocommand for Vim that would only be triggered when the popup menu is visible, but I thought Neovim already has this...

**Comment by TheZoq2** [comment]
`:help TextChangedP` does not show anything in nvim 0.2.2

**Comment by Shougo** [comment]
@TheZoq2 It is not implemented.  The patch is available for Vim8 only.

**Comment by Shougo** [comment]
I will port the PR to neovim after merging.
It is needed for deoplete.

**Comment by prabirshrestha** [comment]
@Shougo `TextChangedP` is now in vim 8 master. Would love to have it in neovim soon. https://github.com/vim/vim/commit/5a093437199001a0d60d8e18e2b9539b99a7757c

**Comment by Shougo** [comment]
OK.  Please wait.

**Comment by justinmk** [comment]
TextChangedP merged in #8377
